### ISSUE TITLE
Outgoing SMTP always using DEFAULT account

### ISSUE THREAD

**Issue by MRothaus** [RealRaven2000/SmartTemplate4#51]
This was originally reported on Bugzilla Mozdev Bug #26759 (that was the link when clicking on "New Bug within SmartTemplate4 settings within Thunderbird).  I believe this bug was introduced in SmartTemplate4 v2.8:

This one took me a while to isolate:

I'm using Thunderbird version 68.4.2 on Windows 10 Enterprise.

I have several email accounts configured in Thunderbird.  In Thunderbird, one
of the accounts is defined as "DEFAULT" (the default account is visible via
"Tools / Account Setting").

When I compose a new message, and I use a different account than the default
(by selecting the different account in the FROM line), the message gets sent
using the default SMTP server and credentials.  This means the recipient of the
email is seeing my default account username instead of the chosen account
username...Thunderbird is basically ignoring the FROM address and always using
the default account FROM address.

I did some googling, and somebody had a similar problem, and the solution was to
disable all add-ons, which fixed his problem.

Unfortunately for me, I narrowed it down and disabled SmartTemplate4, restarted
Thunderbird, composed an email with a non-default account, and it worked as
expected.  The recipient sees the FROM as I intended.

So I completely uninstalled SmartTemplate4, restarted Thunderbird, and
installed SmartTemplate4 (which looks like it was updated Feb 7, today).  I did
absolutely no configuring of SmartTemplate4, and sent a message from a
non-default account.  The bug is still there, the FROM is being ignored.

Can you reproduce this issue and hopefully fix it?

I am a registered SmartTemplate4 customer (but I used mrothaus@gmail.com when I
registered).

**Comment by RealRaven2000** [comment]
> So I completely uninstalled SmartTemplate4, restarted Thunderbird, and
> installed SmartTemplate4 (which looks like it was updated Feb 7, today). I did
> absolutely no configuring of SmartTemplate4, and sent a message from a
> non-default account. The bug is still there, the FROM is being ignored.

If you had SmartTemplate configured beforehand then uninstalling is _not going to reset_ your settings. So if you had something configured it will come back. As far as the expected behavior goes, SmartTempalte does not actually change any of your Email headers, unless you use any of the `%header.set%` variables. So here are some questions:

- what template settings to you set when replying? (you can use the save template feature to export for each account)
![image](https://user-images.githubusercontent.com/6314674/74087289-9a1cbd00-4a82-11ea-823e-9a8a75ca9d11.png)
- do you change the "from" by using the dropdown on top of the Composer window? Or some other way? Do you a select a non-default account before you write the email? (by selecting a folder or account in the folder tree)? Or is there a third way I do not know of.

**Comment by MRothaus** [comment]
After I choose "Write" from main Thunderbird window, I change the "FROM" using the drop-list in the new message window, if I want to use a non-default account (the default account is already in the FROM field).
I do find it odd that somehow SmartTemplate4 is causing Thunderbird to essentially ignore the account chosen in the FROM field, but that is behavior I'm witnessing.  When I disable SmartTemplate, the FROM non-default account is not being ignored.
The template settings are irrelevant, because I am creating a new message, not replying or forwarding an existing message.  But I am attaching the saved template for your review.
    On Saturday, February 8, 2020, 09:54:07 AM EST, RealRaven2000 <notifications@github.com> wrote:  
 
 

So I completely uninstalled SmartTemplate4, restarted Thunderbird, and
installed SmartTemplate4 (which looks like it was updated Feb 7, today). I did
absolutely no configuring of SmartTemplate4, and sent a message from a
non-default account. The bug is still there, the FROM is being ignored.


If you had SmartTemplate configured beforehand then uninstalling is not going to reset your settings. So if you had something configured it will come back. As far as the expected behavior goes, SmartTempalte does not actually change any of your Email headers, unless you use any of the %header.set% variables. So here are some questions:
   
   - what template settings to you set when replying? (you can use the save template feature to export for each account)   

   - do you change the "from" by using the dropdown on top of the Composer window? Or some other way? Do you a select a non-default account before you write the email? (by selecting a folder or account in the folder tree)? Or is there a third way I do not know of.

**Comment by MRothaus** [comment]
The common.json looks like it didn't get attached and it isn't supported. So here is the saved template text:

{
  "newmsg": "",
  "rspheader": "",
  "rspmsg": "\n-------- Original Message --------\nSubject: %subject%\nDate: %X:=sent%%a%, %e% %b% %Y%, %H%:%M% [UTC%date_tz%]\nFrom: %from(name,bracketMail())%\nTo: %to(name,bracketMail())%[[\nCC: %cc(name,bracketMail())%]]\n",
  "fwdheader": "",
  "fwdmsg": "",
  "new": false,
  "newhtml": false,
  "newnbr": true,
  "rsp": true,
  "rsphead": true,
  "rsphtml": false,
  "rspnbr": true,
  "fwd": false,
  "fwdhead": false,
  "fwdhtml": false,
  "fwdnbr": true
}

**Comment by RealRaven2000** [comment]
Looks pretty standard. Do you see exceptions in error console when you switch?

**Comment by RealRaven2000** [comment]
Very strange -when I send and change to my hotmail account it looks ok on the sender side:

![image](https://user-images.githubusercontent.com/6314674/74094951-5907c580-4ae1-11ea-9ad0-d82707899d22.png)

receiving it looks lik ethis: same username but email was relayed via gmail so it looks like it was from my first account (axel ..@gmail) but with the **user name** of the correct account I selected (from: Realraven)

![image](https://user-images.githubusercontent.com/6314674/74094964-818fbf80-4ae1-11ea-826d-ad1031331923.png)

**Comment by MRothaus** [comment]
I've always seen that behavior, where the username is correct, but the primary account is included when you view the message source.

I didn't see any exceptions in the error console.

I think I'll try a completely fresh install of TBird and see if that makes this problem go away.  I do have several different email accounts, but they all using google mail for processing (I have a personal GMail account, and 2 other paid domains that use google for mail processing).

**Comment by MRothaus** [comment]
I just turned on the error console again, left it up, did a "Write" to compose a new message, and I got these 2 errors:

NS_NOINTERFACE: Component returned failure code: 0x80004002 (NS_NOINTERFACE) [nsISupports.QueryInterface] stack-trace-collector.js:84
TypeError: this.NotifyComposeBodyReadyNew is not a function MsgComposeCommands.js:381:14

I then changed the "FROM" by choosing a non-default from the drop list, but no more errors occurred in the error console.

**Comment by MRothaus** [comment]
I decided to do a complete uninstall of Thunderbird, and even deleted the AppData\Roaming\Thunderbird directory.  Here are the steps I did afterwards:

- Fresh install of Thunderbird 68.4.2

- configured my google mail personal account (xxx@gmail.com); Thunderbird automatically created a SMTP outgoing settings for me (did a nice job of it too).

- sent a message to my yahoo account -> my yahoo mail received it properly from xxx@gmail.com (which I expected)

- added another account, my work account, (xxx@argentek.com), Thunderbird automatically configured the outgoing SMTP settings for me (it is my own domain, which uses google back-end for email).  I made my work account the default SMTP.

- sent a message to my yahoo, but changed the "FROM" droplist from the default Argentek to my personal gmail -> received properly at yahoo from xxx@gmail.com

- installed SmartTemplate4 extension, and did no configuring at all.  I restarted Thunderbird.

- sent another message to my yahoo, I saw the "SmartTemplate BUY LICENSE" in the compose window, changed the FROM droplist from the default Argentek to my personal gmail -> received at my yahoo account, but the FROM is xxx@argentek.com.  Disappointing.

- I disabled SmartTemplate4, restarted Thunderbird, did same steps changing FROM to my personal gmail and sent to my yahoo -> received properly from xxx@gmail.com at yahoo.

- Finally, I re-enabled SmartTemplate4, sent a message from my yahoo account to xxx@gmail.com, received it in Thunderbird, hit REPLY to send back to my yahoo, and the FROM account at my yahoo is correct xxx@gmail.com.  So replies are working, it is just the new messages are not.

I don't know what could be the underlying problem here.  Maybe because both personal and work accounts use the same smtp.gmail.com outgoing server (but with different usernames and passwords)?  Clearly it is some bad interaction between Thunderbird and SmartTemplate4.

**Comment by MRothaus** [comment]
I remembered I had an "xxx@verizon.net" email account, so I added that account which does not use google back end.  The same problem occurs (changing the default FROM to my xxx.verizon.net shows up as the default account in yahoo email).  So you should be able to reproduce this with 2 different email accounts configured in a fresh thunderbird install with a fresh SmartTemplate4 installed.

**Comment by MRothaus** [comment]
Nobody else is having this problem?  That's surprising.

**Comment by Bishop-Berkley** [comment]
Yes, I am having this issue too, except that my emails are not being sent though the default SMTP, they are being sent through the SMTP associated with the account that received the original email to which I am replying or from which I am forwarding.  The issue exists with replies and forwards, and deleting the original message makes no difference.  The original email account is 'sticky'.  So, to be specific:

Prior to TB 68, you used to be able to receive an email in (say) an old email account's inbox and then reply or forward using (say) your new email account through its SMTP.  But since the 68 update - running SmartTemplate 2.9.1 - if you receive an email in Account 1 (IMAP) and you attempt to reply or to forward that email after choosing a different Account 2, by selecting your Account 2 from the pull down menu in the "From" field in the message window, the message will appear to indicate that it will be sent using Account 2 and its SMPT server, which is set out in parentheses after the account name, but TB will nevertheless throw up an error. I noticed that the Thunder Bird Sending Message dialog box that appears (for about 1 second) after you press the send button states that it is sending via Account 1's SMTP even though you have changed the "From" field. Therefore TB returns the error: "An error occurred while sending mail. The mail server responded: Request failed; Mailbox unavailable. Please verify that your email address is correct in your account settings and try again" because (or course) it is using the Account 1 SMTP but using your Account 2 credentials.

The issue is 100% consistent/reliable, and I am sure it came up with TB 68 because I do this action all the time, and only had this issue after that update.

Disabling SmartTemplate solves this issue with no other changes.

**Comment by Bishop-Berkley** [comment]
btw: I am new to this, but I think the owner of this thread may wish to label this as a "Bug"

**Comment by MRothaus** [comment]
I looked around, and couldn't find a way to add the "Bug" label.  I haven't heard back from RealRaven, so I don't think he has a solution.

**Comment by JoeWinograd** [comment]
Hi Michael,

First, thank you for reporting this bug. I am experiencing it, too. I found your bug report at Bugzilla, but have not been able to sign up there, so I decided to write directly via the "Contact us via e-Mail" button in the ST tab in Thunderbird. The reply from Axel pointed me to this GitHub interface and your report in it. For the sake of completeness here, I'll enter my email thread with Axel:

----- Begin email thread with Axel -----

--- From Joe:

W10 Pro V1909 64-bit
Thunderbird 68.6.0 64-bit
SmartTemplate4 Pro 2.9.1 one-year license valid until 08-01-20

I've tried but have been unable to create an account at Bugzilla (it never sends the confirmation email), so I decided to send you this email directly.

I have many email accounts configured in Thunderbird, some with the same SMTP sever, some with different ones. There is a serious bug in SmartTemplate4 that causes Thunderbird to use the wrong SMTP server and email account when the account is not the default. This has already been reported by Michael Rothaus (xxxxx@gmail.com) on 2-Feb-2020:

https://www.mozdev.org/bugs/show_bug.cgi?id=26759

As that was almost six weeks ago, I wanted to add my emphasis on how severe a bug this is. In fact, it makes SmartTemplate4 unusable from my perspective. Yes, I could change the default account every time I compose an email (when it's not already the default account), but, quite frankly, that's untenable, as I switch accounts frequently.

I see that Michael set the severity at Major rather than Critical because of that work-around, but I think he's being too generous. This is a **Critical** bug, imo, that warrants an immediate fix, yet it's already been nearly six weeks. I urge you to fix this right away. Thank you for your attention to this matter.

--- From Axel:

for bug reports, you can use the new github interface (preferred) here:

https://github.com/RealRaven2000/SmartTemplate4/issues

I think it is this issue:

https://github.com/RealRaven2000/SmartTemplate4/issues/51

the question is - iffyou disable SmartTemplte templates in the ST4 account dropdown in common and on the actual mail account (but not disable ST4 Add-on altogether) does this still occur? It's hard to understand as SmartTemplate does not/ shouldn't interfere with the "from" address unless you use %header.set(from)%.

--- From Joe:

Thanks for the fast reply...much appreciated! And thanks for the link to the new github interface...I'll sign up now.

> I think it is this issue:

Yes, that is the issue! It is 100% reproducible here. If I disable ST4, sending works correctly, i.e., Thunderbird uses the right SMTP server and the right email account.

> the question is - if you disable SmartTemplte templates in the ST4 account dropdown in common and on the actual mail account (but not disable ST4 Add-on altogether) does this still occur?

I don't know. To prove that ST4 is the culprit, I disabled it, which did prove conclusively that ST4 causes the problem.

> It's hard to understand as SmartTemplate does not/ shouldn't interfere with the "from" address unless you use %header.set(from)%.

I don't use that. My **Reply** and **Forward** templates are identical...and very simple...here it is:

**-------- Original Message --------
Subject: %subject%
Date: %X:=sent% %a% %d%-%b%-%Y% %H%:%M%:%S% %date_tz%(%tz_name%)
From: <%from(mail,link)%>
To: <%to(mail,link)%>
[[Cc: <%cc(mail,link)%>]]**

Are you saying that you can't reproduce it? If so, that really surprises me. Thanks for your help.

----- End email thread with Axel -----

To answer the questions above from RealRaven2000:

Q: what template settings to you set when replying?

A: Reply and Forward templates are shown above.

Q: do you change the "from" by using the dropdown on top of the Composer window?

A: Yes.

Q: Do you a select a non-default account before you write the email? (by selecting a folder or account in the folder tree)?

Account selection is via the **From** drop-down.

Q: do you use any %header% variables?

A: No. The complete templates are shown above.

Btw, Michael's work-around (thanks for that!) works here, too, but, as I noted in my email to Axel, having to change the default account every time you compose an email (when it's not already the default account) is untenable and makes ST unusable, imo.

Regards, Joe

**Comment by MRothaus** [comment]
I have a hunch that the "SmartTemplate BUY LICENSE" message inserted by SmartTemplate on messages is inadvertently affecting the "From" droplist somehow.  I leave SmartTemplate enabled when I REPLY to emails, because I normally don't need to touch the "From" droplist when replying, the default FROM in the reply message is what I want anyway.

But on a new message, when I need to change from the default account in the "FROM" droplist, I know it won't work.  Either I have to disable SmartTemplate before creating a new message, or I find an old message from the account I want to send from, and reply to that message.  Changing the "To" field is no problem (it is a textbox field, not a droplist).  It is only the FROM droplist that does not respect a change to it when SmartTemplate is enabled.

Like I said earlier, this is reproducible with a fresh install of TBird and a fresh install of SmartTemplate and 2 different email accounts.

**Comment by JoeWinograd** [comment]
> the "SmartTemplate BUY LICENSE" advertisement inserted by SmartTemplate on messages

I've never seen that (probably because I purchased a one-year license).

> I normally don't need to touch the "From" droplist when replying, the default FROM in the reply message is what I want anyway.

Same here, but the problem is that even though the **From** field/drop-down displays the correct account when doing a Reply (or Forward), **the actual SMTP sever/account selected is wrong**. Regards, Joe

**Comment by RealRaven2000** [comment]
> > I normally don't need to touch the "From" droplist when replying, the default FROM in the reply message is what I want anyway.
> 
> Same here, but the problem is that even though the **From** field/drop-down displays the correct account when doing a Reply (or Forward), **the actual SMTP sever/account selected is wrong**. Regards, Joe

FWIF, that one I can reproduce here. I really need to debug Thunderbird itself to see what it does when ST4 is disabled.

**Comment by JoeWinograd** [comment]
> that one I can reproduce here

Great news! That's a huge step in the right direction. Thanks for the update...much appreciated! Regards, Joe

**Comment by MRothaus** [comment]
> I've never seen that (probably because I purchased a one-year license).

I purchased the license too.  I saw the ST advertisement when I was reproducing the problem with a fresh TBird install followed by a fresh ST install (to rule out a problem with my specific installation and configuration).
 
> Same here, but the problem is that even though the **From** field/drop-down displays the correct account when doing a Reply (or Forward), **the actual SMTP sever/account selected is wrong**. Regards, Joe

Interesting!  I didn't notice that.  Most likely because my accounts all use google servers.  I'm glad you found this thread and are adding valuable information to help solve the problem.

**Comment by JoeWinograd** [comment]
> Most likely because my accounts all use google servers.

Mine do not, which is why it's such a critical bug for me. I have accounts with AT&T, Comcast, Gmail, my company domain, my personal domain, my college domain, and more. It is a huge problem to have an email be sent from the wrong SMTP server/account.

I was fortunate to notice it the first time it happened because I don't store my account passwords in the TBird master...I enter them manually. So, the first time it occurred I was responding to a business email when the Enter Password dialog came up for my personal SMTP server/account! I was, to say the least, stunned!

Of course, it's even more insidious after that, because after entering an SMTP/account password the first time, TBird never asks again in that session. The sending dialog flashes on the screen, but it's usually too fast to notice that it sent the email from the wrong SMTP server/account! Let's hope that Axel comes up with a fix quickly. Regards, Joe

**Comment by RealRaven2000** [comment]
so my first real test was with 2 real SMTP accounts, thunderbirddaily@gmail.com and thunderbirddailyt67@gmail.com

![image](https://user-images.githubusercontent.com/6314674/77339186-5248b100-6d23-11ea-893f-60c8cf1a1e2c.png)

I started writing the email form the tb67 account and then changed the from to the thunderbirddaily one. It's displayed correctly (and also was saved into the correct SENT folder). The question is, does this bug only affect alias Email addresses? Or is it caused by a shared SMTP (Send) Server? What do I need to change on my side to recreate the bug? My guess is that I can only reproduce it with alias ids (using the same email address), so that's going to my next test.

**Comment by RealRaven2000** [comment]
Second test: using an alias identity (with a fake email address) sent from the tb67 account:

![image](https://user-images.githubusercontent.com/6314674/77339652-04807880-6d24-11ea-97ef-a8b2a05e0da4.png)

what it does: it changes the from "Name" but is still using the main email address of the account:
Support <thunderbirddaily67@gmail.com>

So I guess this should be correct behavior to, as one shouldn't be allowed to "fake" ones email address.

**Comment by JoeWinograd** [comment]
> does this bug only affect alias Email addresses?

No. It affects real addresses.

> Or is it caused by a shared SMTP (Send) Server?

Yes, it sends from the wrong account on the same SMTP server. But it also sends on the wrong account when selecting a different SMTP server.

> What do I need to change on my side to recreate the bug?

You should try some tests with email accounts other than **Gmail.com** accounts.

> My guess is that I can only reproduce it with alias ids (using the same email address), so that's going to my next test.

No. That's not the issue,. The bug that I'm experiencing has nothing whatsoever to do with aliases/identities. It occurs with real email addresses on the same or different SMTP servers.

Regards, Joe

**Comment by MRothaus** [comment]
Forget about email aliases.  You did the correct test, 2 different email accounts.  One of them is designated as the DEFAULT account, so that when you do "Compose", the DEFAULT account is displayed in the FROM droplist.  If you choose the other email account in the FROM droplist and send the email, the recipient will still see it show up as if it came from the DEFAULT account.  Your changing the FROM in the droplist will be ignored.

**Comment by MRothaus** [comment]
Let us know when you can reliably reproduce the issue yourself.  I still am suspicious of the "SmartTemplate BUY LICENSE" advertisement insertion into the new message window...that is a recent addition to your code, and the logical place to be causing an issue on a new message, since SmartTemplate doesn't have to process a new message like it does for a reply/forward message.

**Comment by Bishop-Berkley** [comment]
Again, if it helps in finding/reproducing the bug: When I change the **From** drop down selection in an email reply to a second account (not the one in which the original was received), TB appears to accept the change in the composition window, but nevertheless on hitting Send uses the SMTP server associated with the original email account to send the reply (not necessarily the default SMTP server).  This action fails because TB logs in to that original SMTP server using login details saved for that second account, as is evident from the error message.

**Comment by RealRaven2000** [comment]
> Forget about email aliases. You did the correct test, 2 different email accounts. One of them is designated as the DEFAULT account, so that when you do "Compose", the DEFAULT account is displayed in the FROM droplist. If you choose the other email account in the FROM droplist and send the email, the recipient will still see it show up as if it came from the DEFAULT account. Your changing the FROM in the droplist will be ignored.

Looks like it is harder to reproduce than I thought - I have 2 accounts on my test profile, not sure which one is the "default":
1:
![image](https://user-images.githubusercontent.com/6314674/77345669-de130b00-6d2c-11ea-85bf-cf84ffb8dd09.png)

2:
![image](https://user-images.githubusercontent.com/6314674/77345698-e8350980-6d2c-11ea-932c-cb56030cbd33.png)

is it the one that's shown on top?
They both have a default identity.

**Comment by JoeWinograd** [comment]
> I still am suspicious of the "SmartTemplate BUY LICENSE" advertisement insertion into the new message window

I don't see that advertisement when composing a message.

For privacy/security reasons here on the Internet, the email addresses below are not the actual ones, but the actual ones are real email addresses, not aliases/identities.

As an example, the bug occurred on the first email that I tried to send today. The email came into this account:

George@MyPersonalDomain.com

I hit the From drop-down and selected a different account on the same SMTP server:

Henry@MyPersonalDomain.com

I forwarded the email, but when I hit Send, the SMTP password dialog that came up was for:

George@MyPersonalDomain.com

I cancelled out of the password dialog so that the email was not sent.

What's strange about this occurrence of the bug is that the Default at the time was Henry@MyPersonalDomain.com — normally, Michael's work-around of setting the desired account as the Default works, but not this time. Now here's where it gets even stranger. I went into Account Settings, de-selected Henry@MyPersonalDomain.com as the Default, then re-selected it as the Default, then tried to forward the email again, and this time it worked, i.e., the SMTP password dialog that came up was for:

Henry@MyPersonalDomain.com

I entered the password and it was sent via the correct account (on the correct SMTP server). Regards, Joe

**Comment by RealRaven2000** [comment]
> Let us know when you can reliably reproduce the issue yourself. I still am suspicious of the "SmartTemplate BUY LICENSE" advertisement insertion into the new message window...that is a recent addition to your code, and the logical place to be causing an issue on a new message, since SmartTemplate doesn't have to process a new message like it does for a reply/forward message.

There were many more changes and the code is also executed on a "write" template. I have sent you a license to your reliable address (hope that one is still in use) for testing. I have a feeling it has to do with older Thunderbird profiles where accounts were renamed / removed. I will test from my production profile instead maybe it is easier to reproduce there...

**Comment by JoeWinograd** [comment]
> This action fails because TB logs in to that original SMTP server using login details saved for that second account, as is evident from the error message.

**What's really dangerous here is that it does not fail!** It sends the email from the wrong account on the wrong SMTP server, but if those credentials have already been entered in this TBird session, the send works — using the wrong (but valid) account/SMTP server. This is why it is such a critical, egregious bug.

**Comment by RealRaven2000** [comment]
> There were many more changes and the code is also executed on a "write" template. I have sent you a license to your reliable address (hope that one is still in use) for testing. I have a feeling it has to do with older Thunderbird profiles where accounts were renamed / removed. I will test from my production profile instead maybe it is easier to reproduce there...

maybe in your case the replying from the original account is the key?

**Comment by RealRaven2000** [comment]
> > This action fails because TB logs in to that original SMTP server using login details saved for that second account, as is evident from the error message.
> 
> **What's really dangerous here is that it does not fail!** It sends the email from the wrong account on the wrong SMTP server, but if those credentials have already been entered in this TBird session, the send works — using the wrong (but valid) account/SMTP server. This is why it is such a critical, egregious bug.

That's why it would be a good idea to see whether (first SmartTemplate debug mode disabled, then maybe later with it enabled) there are any additional message in JavaScript error console **when you switch over**.

I am especially interested in any exceptions / warnings that are shown during this, Bring it up with Ctrl+Shift+J (from the main window) while the composer window is open, **clear the console window content before you change** the "from" address. 
![image](https://user-images.githubusercontent.com/6314674/77348236-01d85000-6d31-11ea-9f46-f11c4f9f3cfb.png)


Also it is a MAJOR detail for the bug whether you have _already started typing_ (before selecint a different _from_) or not. After you started typing ST4 will not attempt running the composer code a second time (because this could cause loss of data / what you have typed). This changes the behavior much more than any license warnings. I can also send you an unofficial way of extending the trial period (in email not in this issue) if you like.

Having said that, neither modes overwrite the original / reply header (unless you use `%header.(..)%` functions)

**Comment by JoeWinograd** [comment]
> not sure which one is the "default"

The underlined one is the Default.

> is it the one that's shown on top?

When you set a different one as the Default, it stays where it is in the order during that opening of Account Settings, but it becomes underlined. When you click OK to exit Account Settings and then go back into Account Settings, it will be at the top (of course, still underlined).

> But they are both from google (gmail), so maybe going across completely different ISPs causes the issue?

As I mentioned earlier, it happens here in both cases, i.e., different accounts on the same SMTP server and different accounts on different SMTP servers. Regards, Joe

**Comment by Bishop-Berkley** [comment]
> > This action fails because TB logs in to that original SMTP server using login details saved for that second account, as is evident from the error message.
> 
> **What's really dangerous here is that it does not fail!** It sends the email from the wrong account on the wrong SMTP server, but if those credentials have already been entered in this TBird session, the send works — using the wrong (but valid) account/SMTP server. This is why it is such a critical, egregious bug.

I have not experienced that. In my forward/reply scenario the original SMTP (associated with the account which received the original message) is sent the outgoing message with login/password credentials I have saved for the SMTP server I am intending to use.  So it returns a "Request failed; Mailbox unavailable. Please verify that your email address is correct in your account settings and try again" error.

**Comment by RealRaven2000** [comment]
Ok, I am one step further - I can "in some cases" reproduce this on my production profile. If I type something first and then change the "from" address the message is sent of as expected. However if (in my gmail POP3 account) I change the from _before I start typing_, it appears to change the from address correctly on the send side (and even saves the message with the correct from; I am not sure if it would put it into the SENT folder as I am using "copy sent to current" - when it arrives on the other side it still shows the default (gmail) account.

So factually Thunderbird does something in both cases but in the second case SmartTemplate does additional work (running the processing with the new identity) and maybe it does it too early or stops Thunderbird completely doing what it is supposed to do. The problem with that is that I need to debug "whatever Thunderbird is doing when it changes the address" **completely** plus I need to understand all the steps it does then. Which is a tall order...

**Comment by MRothaus** [comment]
> There were many more changes and the code is also executed on a "write" template. I have sent you a license to your reliable address (hope that one is still in use) for testing. I have a feeling it has to do with older Thunderbird profiles where accounts were renamed / removed. I will test from my production profile instead maybe it is easier to reproduce there...

FWIW, I did a totally fresh install of TBird followed by a fresh install of ST (and not applying the license), and I was able to reproduce the bug.  I made sure I had no leftover files after I uninstalled TBird before doing the test (e.g. I deleted the folder holding the TBird profiles).

**Comment by JoeWinograd** [comment]
> Also it is a MAJOR detail for the bug whether you have already started typing (before selecint a different from) or not.

I have not. In fact, I write all emails in my fav text editor and then copy/paste into the TBird compose window (I also do that for Internet posts...am typing right now in my text editor...will come your way via a copy/paste into the comment block).

> I can also send you an unofficial way of extending the trial period (in email not in this issue) if you like.

Not an issue here. As I mentioned in [my first post ](https://github.com/RealRaven2000/SmartTemplate4/issues/51#issuecomment-601728982), I have SmartTemplate4 Pro 2.9.1 with a one-year license valid until 1-Aug-2020.

> unless you use %header.(..)% functions

As mentioned earlier, I do not use those functions (I posted my entire Reply and Forward templates in [my first post ](https://github.com/RealRaven2000/SmartTemplate4/issues/51#issuecomment-601728982). Regards, Joe

**Comment by RealRaven2000** [comment]
The licensing stuff is a red herring, let's drop that. It's to do with the internal Thunderbird function "LoadIdentity()" - which I am replacing and calling manually. Can you try the test version below, just rename from .zip to .xpi (github doesn't allow uploading xpi files) - then drag it into your Add-ons manager, install, restart, test.

[smartTemplate-fx-2.9.2pre26.zip](https://github.com/RealRaven2000/SmartTemplate4/files/4371322/smartTemplate-fx-2.9.2pre26.zip)

**Comment by MRothaus** [comment]
I installed the xpi, and the results are very promising!  I did several tests to my yahoo email from TBird, and the FROM droplist is being respected.  I also sent from an email alias, and it used the correct SMTP server this time, but the Yahoo email system shows the email coming from the main account, not the alias....but that has been happening for a while now and nothing ST can do about it.

I hereby officially retract any suspicions of the licensing advertisement insertion. :)

**Comment by RealRaven2000** [comment]
> I installed the xpi, and the results are very promising! I did several tests to my yahoo email from TBird, and the FROM droplist is being respected. I also sent from an email alias, and it used the correct SMTP server this time, but the Yahoo email system shows the email coming from the main account, not the alias....but that has been happening for a while now and nothing ST can do about it.

If we look at the main difference in code, here is what happens:

SmartTemplate will overwrite the Tb function LoadIdentity with it's own "loadIdentity" in order to catch the event when the Identity is set (when Composer starts) or changed (when you drop down the "FROM" address). It keeps a reference to the original function which it calls in the following cases:

1 - on startup. This gives thunderbird a chance to do its own processing (which usually includes inserting the quote when replying and "on ... Joe wrote:" - the original quote header. SmartTemplate doesn't do anything else in this case and waits for the `NotifyComposeBodyReady` event, where it executes its payload

2 - when the "FROM" is selected: it checks whether the body of the mail was modified (if you have typed / edited / deleted anything)
2a) if the email is unchanged, SmartTemplate calls its `insertTemplate` function
2b) if the mail was changed, the signature may still be processed / replaced. Also, in a new email all "deferred variables" (the ones that you can click on to replace, which have some JavaScript attached to them) have to be replaced, because the JavaScript still assumes the old identity. If no template stuff was processed (ST4 template not active on this account) only then the original LoadIdentity function was called. THis basically lets Thunderbird do its own Composition preparation (see above)

So, in the case (2a)  the Thunderbird function `LoadIdentity` was not called. It was regarded as superfluous, because ST4 does it's own processing based on what's on screen. Little did I know that there is probably some back-end code that rewrites the "from" header (and not just changes it on screen). In the pre26 patch I added the call to the original Load if the body is not manually modified by the user, before calling its own `insertTemplate(..)` function:

```
loadIdentity: function loadIdentity(startup, previousIdentity) {
  ...

  if (!isBodyModified) {
    // [issue 51]
    this.original_LoadIdentity(false); // make sure Tb does everything it needs to the from header!
    // Add template message - will also remove previous template and quoteHeader.
    this.smartTemplate.insertTemplate(false);
    // [Bug 25104] when switching identity, old sig does not get removed.
    //             (I think what really happens is that it is inserted twice)
    isTemplateProcessed = true;
  }
```

**Comment by JoeWinograd** [comment]
> Can you try the test version below, just rename from .zip to .xpi (github doesn't allow uploading xpi files) - then drag it into your Add-ons manager, install, restart, test.

Should we remove 2.9.1 first? Will 2.9.2 retain our Pro license?

**Comment by MRothaus** [comment]
Joe, I just installed the new xpi over the old one, didn't remove it first.  My license was retained.

Axel: I think 99% of the time when I compose a new message, I change the FROM first (case 2a), then compose my message.  In my testing today, I did both cases 2a and 2b and both are working properly with your patched code.

By the way, I did another test using an email alias, but with ST4 disabled.  I sent the email to my yahoo account and it got received and displayed with the main account...identical behavior with ST4 enabled or disabled.  So as far as I'm concerned, ST4 patch is working for me.  Joe has a more complicated configuration and I hope he has the same success as I do.

**Comment by RealRaven2000** [comment]
> 
> 
> > Can you try the test version below, just rename from .zip to .xpi (github doesn't allow uploading xpi files) - then drag it into your Add-ons manager, install, restart, test.
> 
> Should we remove 2.9.1 first? 
Nope. Most likely this patch is going into the release version, like almost every other one.

> Will 2.9.2 retain our Pro license?
Of course! It's still the same Add-on, just a prerelease.

**Comment by JoeWinograd** [comment]
Tested here with many scenarios...all worked perfectly! I'm convinced that you found the bug...can't thank you enough, Axel, for fixing this...awesome!

My thanks, also, to Michael and Bishop for their continuing participation in this thread that helped lead to a solution, with a special shout-out to Michael for the initial bug report and the work-around.

My appreciation one more time to Axel! Regards, Joe

**Comment by MRothaus** [comment]
Great news, Joe!  Axel, I'll leave the honor to you to close this bug report at your convenience.

**Comment by RealRaven2000** [comment]
> Great news, Joe! Axel, I'll leave the honor to you to close this bug report at your convenience.

thanks - good news indeed. I usually close my bugs when I release officially and everyone gets to enjoy the fix, to avoid confusion. Updated version history:
http://smarttemplate4.mozdev.org/version.html#2.9.2

**Comment by Bishop-Berkley** [comment]
Thanks, I will test tonight and report back.

**Comment by Bishop-Berkley** [comment]
> 
> 
> Thanks, I will test tonight and report back.

smartTemplate-fx-2.9.2pre26 works fine for me.  It is not exhibiting the bug in circumstance where it once was consistently.

**Comment by RealRaven2000** [comment]
Fixed in 2.10.1 released on 14/04/2020
